### PR TITLE
[mypyc] Add cleanup at end of init function

### PR DESCRIPTION
**Pull request by pranavrajpal**
This frees everything created in the generated initialization function so that we don't leak memory in the case of an error during initialization, as mentioned in https://github.com/python/mypy/pull/10586#pullrequestreview-678090095.

### CODE REVIEW

File: mypyc/codegen/emitmodule.py [hunk 1]
```diff
@@ -899,9 +899,11 @@ def generate_module_def(self, emitter: Emitter, module_name: str, module: Module
                            '    goto fail;')
 
         # HACK: Manually instantiate generated classes here
+        type_structs = []  # type: List[str]
         for cl in module.classes:
             if cl.is_generated:
                 type_struct = emitter.type_struct_name(cl)
+                type_structs.append(type_struct)
```
Comment (msullivan) [inline mypyc/codegen/emitmodule.py:906]: I think we need to collect the type structs even for the not generated ones. The `is_generated` classes (like those created for nested functions) get initialized here, but most classes get initialized by code we generated as IR, as part of the module initialization function. 

(I can't remember why we don't arrange to do that for the generated ones; obviously I didn't feel great about that fact, which is why this block of code is labeled `HACK` :))

File: mypyc/codegen/emitmodule.py [hunk 2]
```diff
@@ -918,7 +920,13 @@ def generate_module_def(self, emitter: Emitter, module_name: str, module: Module
 
         emitter.emit_line('return {};'.format(module_static))
         emitter.emit_lines('fail:',
-                           '{} = NULL;'.format(module_static),
+                           'Py_XDECREF({});'.format(module_static),
```
Comment (msullivan) [inline mypyc/codegen/emitmodule.py:923]: Do `Py_CLEAR` for these, which will do an Py_XDECREF and NULL out the variable. (By virtue of being a macro that does not behave like a function...)

File: mypyc/codegen/emitmodule.py [hunk 3]
```diff
@@ -918,7 +920,13 @@ def generate_module_def(self, emitter: Emitter, module_name: str, module: Module
...
+                           'Py_XDECREF(modname);')
+        # the type objects returned from CPyType_FromTemplate are all new references
+        # so we have to decref them
+        for t in type_structs:
```
Comment (msullivan) [inline mypyc/codegen/emitmodule.py:927]: We should also do the finals, too. Those need to be reset to whatever the proper error value for the type is.

Comment (msullivan) [reply mypyc/codegen/emitmodule.py:934]: I think the right thing to do is to call out to `emit_dec_ref`, which knows how to decref everything: https://github.com/python/mypy/blob/master/mypyc/codegen/emit.py#L344

And then assign the result of `c_undefined_value` to the variable, to clear it out

File: mypyc/codegen/emitmodule.py [hunk 4]
```diff
@@ -918,8 +921,19 @@ def generate_module_def(self, emitter: Emitter, module_name: str, module: Module
 
         emitter.emit_line('return {};'.format(module_static))
         emitter.emit_lines('fail:',
-                           '{} = NULL;'.format(module_static),
-                           'return NULL;')
+                           'Py_CLEAR({});'.format(module_static),
+                           'Py_CLEAR(modname);')
+        # the type objects returned from CPyType_FromTemplate are all new references
```
Comment (msullivan) [inline mypyc/codegen/emitmodule.py:926]: I think this comment has drifted away from the code it is describing

Comment (pranavrajpal) [reply mypyc/codegen/emitmodule.py]: This should be fixed in ff8d287f196b98cc55809ecb72a4f4b674bc09bc.

File: mypyc/codegen/emitmodule.py (hunk 3 above)
Comment (pranavrajpal) [reply mypyc/codegen/emitmodule.py:933]: Yeah, that does look a lot cleaner. This should be fixed in ebe20ed7d03802584fdd593ac77224e9be3f8022.

**Pull request merged**